cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -748,18 +748,77 @@
 (decl put_in_reg_mem (Value) RegMem)
 (extern constructor put_in_reg_mem put_in_reg_mem)
 
+;; Addressing modes.
+
 (type SyntheticAmode extern (enum))
 
 (decl synthetic_amode_to_reg_mem (SyntheticAmode) RegMem)
 (extern constructor synthetic_amode_to_reg_mem synthetic_amode_to_reg_mem)
 
+(decl amode_to_synthetic_amode (Amode) SyntheticAmode)
+(extern constructor amode_to_synthetic_amode amode_to_synthetic_amode)
+
 (type Amode extern (enum))
 
+(decl amode_imm_reg (u32 Gpr) Amode)
+(extern constructor amode_imm_reg amode_imm_reg)
+
 (decl amode_imm_reg_reg_shift (u32 Gpr Gpr u8) Amode)
 (extern constructor amode_imm_reg_reg_shift amode_imm_reg_reg_shift)
 
-(decl amode_to_synthetic_amode (Amode) SyntheticAmode)
-(extern constructor amode_to_synthetic_amode amode_to_synthetic_amode)
+;; A helper to check if a shift amount (the `Value`) is both constant and
+;; less-than or equal to 3; this is needed since x64 can only shift addresses
+;; using two bits.
+(decl small_const_shift (u8) Value)
+(extern extractor small_const_shift small_const_shift)
+
+;; A helper to check if a constant offset value will fit into the 32-bit maximum
+;; allowed by the x64 SIB encoding. (TODO re-document, this might not be the
+;; correct explanation)
```

Comment:
From the implementation below it looks like it's actually checking whether the value will fit in a `u32`, so basically, positive values <= 4GiB? Is that right? Do we want to check <= 2GiB or -2GiB <= x <= 2GiB instead?

(pre-existing confusion I know but it's worthwhile to fix this while we're here!)

Reply:
Good catch; that logic was all wrong. Take a look at the new helpers and their logic.